base/loading.jl

```diff
@@ -859,3 +859,40 @@ macro __DIR__()
     __source__.file === nothing && return nothing
     return abspath(dirname(String(__source__.file)))
 end
+
+
+function propagate_sourceloc(sourceloc, ex)
+    if ex isa Expr
+        if ex.head == :macrocall
+            ex.args[2] = sourceloc
+        else
+            map!(e->propagate_sourceloc(sourceloc,e), ex.args, ex.args)
+        end
+    end
+    ex
+end
+
+"""
+    @propagate_sourceloc(quoted_block)
+
+Make all macro calls inside `quoted_block` see the `__source__` location of
+the parent macro call.  In particular, this makes the `@__LINE__` and
+`@__FILE__` macros return the location where the parent macro was invoked.
+For example:
+
+```
+macro example()
+    @propagate_sourceloc quote
+        @__LINE__
+    end
+end
+
+@example
+```
+
+This produces line 7 (the place where `@example` was called) rather than line
+3 (the location of the `@__LINE__` invocation).
+"""
+macro propagate_sourceloc(ex)
+    :(propagate_sourceloc($(esc(:__source__)), $(esc(ex))))
+end
```

Comment:
Sure. How will we deal with 0.6 compatibility?

For context, I ported this utility from here: https://github.com/c42f/MicroLogging.jl/blob/e37dffcb7aeb9c581a27b897fa9d4fe3467d8b85/src/util.jl#L34 where I'm trying to support both 0.6 and 0.7; in 0.6 the `__source__` variable isn't present which is the reason for the macro.

What's the sanest way to deal with getting a nop version of this into `Compat`?  I could add something to make the following work in 0.6, by just removing the function call.  A bit messy!

```julia
@compat propagate_sourceloc(__source__, quote
    @__LINE__
end)
```

Alternatively I could just move the `@propagate_sourceloc` macro to `Compat`...